README.md
# Commerce Layer React Utils

This is a simple collection of React utils and external config files to be used in Commerce Layer micro front-ends.

## How to install

```
npm install @commercelayer/react-utils
```

or

```
yarn add @commercelayer/react-utils
```

## How to use

### Tailwind config

Tailwind default configs can be imported directly into your local `tailwind.config.js` file

```js
const config = require("@commercelayer/react-utils/configs/tailwind")

module.exports = {
  ...config,
}
```

In this way it's possibile to extend the configuration my merging or spreading your own setup.

Example:

```js
module.exports = {
  ...config,
  theme: {
    ...config.theme,
    fontFamily: {
      sans: ["Your custom font"],
    },
}
```

### GlobalStylesProvider

This component takes a primary color HEX string as prop, convert it in HSL value and inject (using styled-components) global css custom properties at `:root` devel.

These custom properties includes varation of the primary color (`--primary-light` and `--primary-dark`), along with a special `--contrast` color to be used inside buttons or element with primary background.

Example:

```jsx
import { GlobalStylesProvider } from "@commercelayer/react-utils"

const App = () => {
  return (
    <GlobalStylesProvider primaryColor="#fff000">
      <YourApp />
    </GlobalStylesProvider>
  )
}
```

We also expose a react component that doesn't use `styled-components`, but directly add css in the dom.

Example:

```jsx
import { InjectCssCustomProperties } from "@commercelayer/react-utils"

const App = () => {
  return (
    <>
      <InjectCssCustomProperties primaryColor="#fff000" />
      <YourApp />
    </>
  )
}
```

### Commerce Layer Logo

This is a simple SVG logo exported as React component.

Example:

```jsx
import { LogoCL } from "@commercelayer/react-utils"

export const Footer = () => {
  return (
    <div>
      Powered by <LogoCL width="135" height="22" className="pl-2" />
    </div>
  )
}
```

### Configure semantic-release in your project

It is possibile to use a sharable semantic-release configuration file, just by creating a new `release.config.js` with the following content

```
module.exports = {
  extends: "@commercelayer/react-utils/configs/semantic-release.js",
}
```

There is no need to add semantic-release packages to your repository, since they are already included as depedencies here.


CHANGELOG.md
# [1.0.0-beta.3](https://github.com/commercelayer/react-utils/compare/v1.0.0-beta.2...v1.0.0-beta.3) (2022-11-21)


### Bug Fixes

* update dependencies ([c156045](https://github.com/commercelayer/react-utils/commit/c156045921d02c15ab1ec2c46bbb80271b333afb))

# [1.0.0-beta.2](https://github.com/commercelayer/react-utils/compare/v1.0.0-beta.1...v1.0.0-beta.2) (2022-06-03)


### Bug Fixes

* add sharable semantic release config ([7990199](https://github.com/commercelayer/react-utils/commit/7990199d6c240b07cb805854605d967957d82c45))

# [1.0.0-alpha.3](https://github.com/commercelayer/react-utils/compare/v1.0.0-alpha.2...v1.0.0-alpha.3) (2022-06-03)


### Bug Fixes

* add sharable semantic release config ([7990199](https://github.com/commercelayer/react-utils/commit/7990199d6c240b07cb805854605d967957d82c45))


src/providers/GlobalStylesProvider/hex2hsl.test.ts
import { hexToHSL, BLACK_COLOR } from "./hex2hsl"

describe("HEX to HSL", () => {
  test("Regular hex code", () => {
    expect(hexToHSL("#ff0000")).toEqual({
      h: 0,
      l: "50.00%",
      s: "100.00%",
    })
  })

  test("Hex that generates floating hue", () => {
    expect(hexToHSL("#ff2a00")).toEqual({
      h: 9.882352941176471,
      l: "50.00%",
      s: "100.00%",
    })
  })

  test("With short hex (3 chars)", () => {
    expect(hexToHSL("#ff0")).toEqual({
      h: 60,
      l: "50.00%",
      s: "100.00%",
    })
  })

  test("With invalid 4 chars hex", () => {
    expect(hexToHSL("#ff00")).toBe(undefined)
  })

  test("With invalid 5 chars hex", () => {
    expect(hexToHSL("#ff001")).toBe(undefined)
  })

  test("With no hash", () => {
    expect(hexToHSL("ff0")).toEqual({
      h: 60,
      l: "50.00%",
      s: "100.00%",
    })
  })

  test("With only hash", () => {
    expect(hexToHSL("######")).toBe(undefined)
  })

  test("With double hash", () => {
    expect(hexToHSL("##ff0")).toEqual({
      h: 60,
      l: "50.00%",
      s: "100.00%",
    })
  })

  test("With spaces", () => {
    expect(hexToHSL(" ##ff0   ")).toEqual({
      h: 60,
      l: "50.00%",
      s: "100.00%",
    })
  })

  test("Empty", () => {
    expect(hexToHSL("")).toBe(undefined)
  })
  test("Empty with spaces", () => {
    expect(hexToHSL("   ")).toBe(undefined)
  })

  test("Broken or invalid hex", () => {
    expect(hexToHSL(" asdasdsasa  ")).toBe(undefined)
  })

  test("Hex with more then 6 chars", () => {
    expect(hexToHSL("#ff2a00222222")).toBe(undefined)
    expect(hexToHSL("ff2a00222222")).toBe(undefined)
  })
})

test("Black color", () => {
  expect(BLACK_COLOR).toEqual({ h: 0, l: "0%", s: "0%" })
})


src/providers/GlobalStylesProvider/hex2hsl.ts
export type HSL = {
  h: number
  s: string
  l: string
}

export const BLACK_COLOR: HSL = {
  h: 0,
  l: "0%",
  s: "0%",
}

export const hexToHSL = (hex: string): HSL | undefined => {
  if (!hex) {
    return undefined
  }

  let hexColor = removeHash(hex.trim())

  if (hexColor.length === 3) {
    hexColor = expand(hexColor)
  }

  const result = /^([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hexColor)
  if (!result) return undefined
  let r = parseInt(result[1], 16)
  let g = parseInt(result[2], 16)
  let b = parseInt(result[3], 16)
  r /= 255
  g /= 255
  b /= 255
  const max = Math.max(r, g, b)
  const min = Math.min(r, g, b)
  let h: number
  let s: number
  const l = (max + min) / 2
  if (max === min) {
    h = s = 0 // achromatic
  } else {
    const d = max - min
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min)
    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0)
        break
      case g:
        h = (b - r) / d + 2
        break
      case b:
        h = (r - g) / d + 4
        break
      default:
        h = 0
        break
    }

    h /= 6
    h *= 360
  }
  const HSL: HSL = {
    h,
    s: `${(s * 100).toFixed(2)}%`,
    l: `${(l * 100).toFixed(2)}%`,
  }

  return HSL
}

const removeHash = (hex: string) => hex.replace(/#/g, "")

const expand = (hex: string) =>
  hex
    .split("")
    .reduce((acc: string[], value) => acc.concat([value, value]), [])
    .join("")


configs/tailwind.js
const colors = require("tailwindcss/colors")
const defaultTheme = require("tailwindcss/defaultTheme")

module.exports = {
  content: [
    "./components/**/*.{js,ts,jsx,tsx}",
    "./pages/**/*.{js,ts,jsx,tsx}",
  ],
  theme: {
    container: {
      screens: {
        sm: "100%",
        md: "100%",
        lg: "100%",
      },
    },
    colors: {
      primary: {
        light: "var(--primary-light)",
        DEFAULT: "var(--primary)",
        dark: "var(--primary-dark)",
      },
      contrast: "var(--contrast)",
      transparent: 'transparent',
      black: '#101111',
      white: '#fff',
      gray: {
        50: "#f8f8f8",
        100: "#EDEEEE",
        200: "#E6E7E7",
        300: "#DBDCDC",
        400: "#878888",
        500: "#686E6E",
        600: "#404141",
        700: "#343535",
        800: "#282929",
        900: "#1D1E1E",
      },
      red: {
        50: "#ffe0e1",
        400: "#FF656B",
        500: "#cc5156",
      },
      green: {
        400: "#1FDA8A",
      },
      orange: {
        400: "#FFAB2E",
      },
    },
    fontFamily: {
      sans: ["Manrope", "ui-sans-serif", "sans-serif"],
    },
    borderColor: theme => ({
      ...theme('colors'),
       DEFAULT: theme('colors.gray.200', 'currentColor'),
     }),
    extend: {
      fontSize: {
        md: "0.938rem",
        ss: "0.813rem",
        xxs: "0.75rem",
      },
      backgroundSize: {
        16: "1rem",
      },
      width: {
        22: "5.75rem",
      },
      minHeight: {
        inherit: "inherit",
      },
      flex: {
        75: "0 0 75px",
        85: "0 0 85px",
      },
      margin: {
        30: "7.5rem",
      },
      boxShadow: {
        bottom: "0 2px 0 0 rgba(0, 0, 0, 0.05)",
        inner: "0 0 0px 1000px #fff inset",
        top: "0px -4px 1px 0px rgb(0, 0, 0, 0.025)",
      },
      transitionProperty: {
        'bg': 'background',
      }
    },
  },
  variants: {
    extend: {
      textColor: ["group-focus"],
      maxHeight: ["group-focus"],
    },
  },
  plugins: [
    require("@tailwindcss/forms")({
      strategy: "class",
    }),
    require("@tailwindcss/line-clamp"),
  ],
}
